prow/cmd/branchprotector/protect.go

```diff
@@ -267,18 +269,94 @@ func (p *protector) UpdateRepo(orgName string, repoName string, repo config.Repo
 		}
 	}
 
+	collaborators, err := p.authorizedCollaborators(orgName, repoName)
+	if err != nil {
+		logrus.Infof("%s/%s: error getting list of collaborators: %v", orgName, repoName, err)
+		return err
+	}
+
+	teams, err := p.authorizedTeams(orgName, repoName)
+	if err != nil {
+		logrus.Infof("%s/%s: error getting list of teams: %v", orgName, repoName, err)
+		return err
+	}
+
 	for bn, githubBranch := range branches {
 		if branch, err := repo.GetBranch(bn); err != nil {
 			return fmt.Errorf("get %s: %v", bn, err)
-		} else if err = p.UpdateBranch(orgName, repoName, bn, *branch, githubBranch.Protected); err != nil {
+		} else if err = p.UpdateBranch(orgName, repoName, bn, *branch, githubBranch.Protected, collaborators, teams); err != nil {
 			return fmt.Errorf("update %s from protected=%t: %v", bn, githubBranch.Protected, err)
 		}
 	}
 	return nil
 }
 
+// authorizedCollaborators returns the list of Logins for users that are
+// authorized to write to a repository.
+func (p *protector) authorizedCollaborators(org, repo string) ([]string, error) {
+	collaborators, err := p.client.ListCollaborators(org, repo)
+	if err != nil {
+		return nil, err
+	}
+	var authorized []string
+	for _, c := range collaborators {
+		if c.Permissions.Admin || c.Permissions.Push {
+			authorized = append(authorized, c.Login)
+		}
+	}
+	return authorized, nil
+}
+
+// authorizedTeams returns the list of slugs for teams that are authorized to
+// write to a repository.
+func (p *protector) authorizedTeams(org, repo string) ([]string, error) {
+	teams, err := p.client.ListRepoTeams(org, repo)
+	if err != nil {
+		return nil, err
+	}
+	var authorized []string
+	for _, t := range teams {
+		if t.Permission == github.RepoPush || t.Permission == github.RepoAdmin {
+			authorized = append(authorized, t.Slug)
+		}
+	}
+	return authorized, nil
+}
+
+func validateRequest(org, repo string, bp *github.BranchProtectionRequest, authorizedCollaborators, authorizedTeams []string) []error {
+	if bp == nil || bp.Restrictions == nil {
+		return nil
+	}
+
+	contains := func(list []string, key string) bool {
+		for _, k := range list {
+			if k == key {
+				return true
+			}
+		}
+		return false
+	}
+
+	var errs []error
+	if bp.Restrictions.Users != nil {
+		for _, c := range *bp.Restrictions.Users {
+			if !contains(authorizedCollaborators, c) {
+				errs = append(errs, fmt.Errorf("Collaborator %s not authorized for %s/%s", c, org, repo))
```

Comment:
Is there a reason why we want to continue after we hit an error here, instead of returning the error right away?

Reply:
Thank you for the review. It is an attempt to capture all the errors, e.g. if there are multiple unauthorised users/teams specified in the `Restrictions` section, rather than reporting only the first error.